### PR TITLE
QuadraticVoteCollectModule voting functionality in process collect added.

### PR DESCRIPTION
**Pull request by MrDeadCe11**
added the quadraticvoteCollectModule voting functionality to the processCollect().  have not put in any signature validation or such.  but the bones are there and ready to go.  minimally tested.  figured I'd put up a PR so it would be too huge.

### CODE REVIEW

File: packages/contracts/contracts/QuadraticFundingCuratorContract.sol
Comment (bitbeckers) [inline packages/contracts/contracts/QuadraticFundingCuratorContract.sol:13]: This isn't a mock, it's a contract/interface that's part of the contract

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:115]: This is risky. Any user can provide data on the round in the encoded data and the checks will pass

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:130]: If possible we probably don't want to use the treasury since that lowers the vote executed. Is this mandatory?

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:217]: Would we need a approval for every executed vote?

File: packages/contracts/test/utils/constants.ts
Comment (bitbeckers) [inline packages/contracts/test/utils/constants.ts:55]: There's a lot of datasets here that are not needed. I'd propose to remove those because it's clutter

File: packages/contracts/test/gitcoinTests/gitcoin.fixture.ts
Comment (bitbeckers) [inline packages/contracts/test/gitcoinTests/gitcoin.fixture.ts:33]: Not sure why you're doing this. All Signers get a starting balance. Iirc this test suite does not fork from any chain?

File: packages/contracts/contracts/TestCollectModule.sol
Comment (bitbeckers) [inline packages/contracts/contracts/TestCollectModule.sol:40]: Not sure, what is the purpose of this Module? If it's for testing the module you're building, you should test on the module you're building. If it's to expose some methods or logging for some reason, it should simply inherit the module you're building.

File: packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol
Comment (bitbeckers) [inline packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol:23]: Why do we need something in the domain of Follow modules?

File: packages/contracts/hardhat.config.ts
Comment (bitbeckers) [inline packages/contracts/hardhat.config.ts:22]: Don't remove this. This check is needed for running deployments. If you require env vars, it's good practice to validate that they are there.

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTests.ts
Comment (bitbeckers) [inline packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTests.ts:65]: Just semantics, but this is not mocked. This is the module we're building if I read the code correct. Labeling it as mocked can become confusing

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts
Comment (bitbeckers) [inline packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts:61]: I think it's cleaner to expect the event that's emitted at the tail end, in this case something like "Vote" and that the parameters of the Vote event match witch what you expect.

**Review by bitbeckers — CHANGES_REQUESTED** [review]
We've already talked bout this, but a few pointers:
* Try to keep PRs smallers because this was a big one to review
* There are confusing and unconsistent things like 'what is being mocked' or 'why is this in the current PR'
* I ran `yarn lint` locally and give a shitton of errors. Please fix those
* Don't remove safety checks.

In general, try to keep your code clean and consistent. It took a lot of back-and-forths between files to figure out was what going on where.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:11]: This needs to be adapted. You don't want to toggle comments in contract code for deployment.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:217]: Yes, that's how it's set up now.  Probably worth re-thinking now that we are looking at implementing claps.

Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:217]: Even without the claps, modules are set with allowance so this shouldn't be part of the voting flow if I get the modules of Lens correctly: https://docs.lens.xyz/docs/approved-allowance-of-modules

File: packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol:23]: The module check some data on this contract.  So I stubbed it to return true.

File: packages/contracts/test/gitcoinTests/gitcoin.fixture.ts
Comment (MrDeadCe11) [reply packages/contracts/test/gitcoinTests/gitcoin.fixture.ts:33]: They get a starting balance of the native token but not of any ERC 20.

Comment (bitbeckers) [reply packages/contracts/test/gitcoinTests/gitcoin.fixture.ts:33]: Ah missed that, thanks

File: packages/contracts/hardhat.config.ts
Comment (MrDeadCe11) [reply packages/contracts/hardhat.config.ts:22]: okay, what if I added a variable to the hardhat.config just a bool that is "useInfuraForTests", that can be true or false?

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTests.ts
Comment (MrDeadCe11) [reply packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTests.ts:65]: ok, I put it as mocked because it's deploying with the mocked lenshub and moduleGlobals contracts.  but yes it is the actual module.  I just wanted to distinguish it in the tests from the the module that was deployed with the actual lens contracts.

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts
Comment (MrDeadCe11) [reply packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts:61]: for some reason the Vote event isn't getting emitted.  if I console.log after the vote event it is emitted but "Vote" is not in the transaction receipt.  that's why I did it like this.  but you're right I should chase down that bug.

File: packages/contracts/contracts/TestCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/TestCollectModule.sol:40]: deleted it.  it was left over something I was trying.  shouldn't have been in the pr.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:11]: ok, I changed this from importing a different contract to mocking the function calls in the tests.

File: packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/mocks/MockFollowValidationModuleBase.sol:23]: deleted this in favor of mocking in the tests.

File: packages/contracts/contracts/QuadraticFundingCuratorContract.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticFundingCuratorContract.sol:13]: moved to interfaces.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:17]: no longer a draft

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:115]: I changed this to get the time stamps from the round module.  but now I wonder if we should store the current active module in the contract and update it whenever a new round is created.  a malicious actor could create a custom round implementation and pass it in as well.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:130]: I think lens is going to want us to put money to their treasury if they are going to incorporate this module.

**Comment by MrDeadCe11** [comment]
comment

**Comment by MrDeadCe11** [comment]
i dunno man, I've reverted back to the exact commit where I closed this and I still can't reopen this.  I'm not sure what else to do.

**Comment by bitbeckers** [comment]
reverted to 779d563 and reopened PR

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:123]: we can add the fee's on the front end and that will move them to be paid by the collector.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol]: Just let the round handle it. If you're calling the round for the limits, and the round checks on limits, skip the first call

Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol]: Check is in round, removed from module

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol]: Approve and vote in same tx

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts
Comment (bitbeckers) [reply packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts]: Fixed and identified blocker.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol]: Fixed

**Review by bitbeckers — DISMISSED** [review]
Merging to unblock.
Cleaned up and added a bunch of TODOs

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:113]: if you're just passing in the currency and amount from the decoded data then this check doesn't do anything.  
[look here for reference](https://github.com/lens-protocol/core/blob/9f3b62aab1a4983d9bc11ee4f5c2892e8f1f148a/contracts/core/modules/FeeModuleBase.sol#L35)
[and here](https://github.com/lens-protocol/core/blob/main/contracts/core/modules/collect/FeeCollectModule.sol#L155)

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:151]: doesn't the vote need to include  [address _token, uint256 _amount,  address _grantAddress,  bytes32 _projectId](https://github.com/gitcoinco/grants-round/blob/5c50f653169f01beb9e69918de3595057ffd8bb3/packages/contracts/contracts/votingStrategy/QuadraticFundingStrategy/QuadraticFundingVotingStrategyImplementation.sol#L62)?

and the project id is new or something because it isn't a parameter in the voting strategy in our resources folder.

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:81]: see other comment on _validateDataisExpected on line 169

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol]: why remove this?  won't this help to keep all the original lens collect features?

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (MrDeadCe11) [inline packages/contracts/contracts/QuadraticVoteCollectModule.sol:25]: I tried to keep all the original data being passed into the profilePublication and just add the addresses we needed.  Is there a reason we don't need to do this?

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts
Comment (MrDeadCe11) [inline packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts:96]: what's going on here?  I thought this worked earlier?

**Review by MrDeadCe11 — COMMENTED** [review]
I see that you're trimming down a lot of the code in the contract (publication init params)  I want to be sure we don't need this to keep continuity with the existing fee collect module.  I guess our module is going to be running along side the existing modules and that's why we don't need the follow validation checks?

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:151]: You're right. Added the grantsRoundAddress and the profileId

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:113]: Well we don't know the amount and arguably you could say the same for the currency except that they need to be accepted as a 'voting currency' in the Grants round

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol:81]: Resolving this to keep discussion focussed

File: packages/contracts/contracts/QuadraticVoteCollectModule.sol
Comment (bitbeckers) [reply packages/contracts/contracts/QuadraticVoteCollectModule.sol]: Do we only allow followers to vote?

File: packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts
Comment (bitbeckers) [reply packages/contracts/test/quadraticFundingCollectionTests/qFundingCollectionTest.unit.ts]: Yes, when the funds were transfered into the module and subsequently routed to the Rounds. This 1) makes the Module the voter and 2) introduces an unintuitive money flow

**Review by bitbeckers — APPROVED** [review]
Merging in the PR and returning to working based of Trello tickets.

**Pull request merged**
